spec.md

```diff
@@ -1957,6 +2319,72 @@ asClock(x)
 [@sec:primitive-operations] will describe the format and semantics of each
 primitive operation.
 
+## Reading Probe References
+
+Probes are read using the `read`{.firrtl} operation.
+
+Read expressions have source flow and can be connected to other components:
+
+```firrtl
+module Foo :
+  output p : Probe<UInt>
+  ; ...
+
+module Bar :
+  output x : UInt
+
+  inst f of Foo
+  x <= read(f.p) ; indirectly access the probed data
+```
+
+Indexing statically (sub-field, sub-index) into a probed value is allowed as
+part of the read:
+
+```firrtl
+module Foo :
+  output p : Probe<{a: UInt, b: UInt}>
+  ; ...
+
+module Bar :
+  output x : UInt
+
+  inst f of Foo
+  x <= read(f.p.b) ; indirectly access the probed data
+```
+
+The source of the probe must reside at or below the point of the `read`
+expression in the design hierarchy.
+
+## Probe
+
+Probe references are generated with probe expressions.
+
+The probe expression creates a reference to a read-only or force-able
+view of the data underlying the specified reference expression.
+
+The type of the produced probe reference is always passive, but
+the probed expression may not be.
+
+Probed memories produce a vector of references to their data.
```

Comment:
Nit: "The type of a probed memory is the type of the memory."

Reply:
Well this is phrased this way because presently in CIRCT we treat it as a FIRRTL vector of data, which implies it's packed.  Sometimes they become unpacked and having a reference claiming to target a packed array leads to incorrect Verilog.  I know you know, but for others, see: https://github.com/llvm/circt/issues/4479 .
One way to look at this is we simply cannot give memories an accurate FIRRTL type presently (certainly not in a way that allows lowering it to both packed and unpacked representations), hence the forced representation as vector of references to the sub-elements.

I do rather like your suggestion language, however!

Reply:
This is pushing up against what I would argue is unspecified behavior. The type of a FIRRTL memory is well defined: `vector<data-type>`. The type of a probe is also well defined `vector<data-type>`. The fact that the unspecified lowering to Verilog produces a mismatch for certain ways that memories seems out-of-scope.

Put differently, in FIRRTL, `vector of references to their data` is equivalent to `the type of the memory`. However, the lowering to Verilog may break this.

I'm not sure how to handle this in the spec except with more details in the compiler implementation section. I also recognize that ref types hit this hard as it is difficult to talk about `read`, `force`, and `release` without bringing in Verilog semantics.

To clarify: I"m fine with the current language.

Reply:
Oh, interesting.  Okay, I get it now (I think), thanks for the explanation!

At risk of somewhat belaboring the point a bit (apologies):

FWIW I don't know that we do have a proper type in FIRRTL for the data within memories?  Spec says what the type of a memory is (bundle of its ports like an instance, which only ever refer to an element at a time, rather indirectly of course), and as an aside this is a great example for why there is probe as separate from just "export a reference to the thing"-- circuit elements are not the same as the data contained within (and in the future it seems likely/useful to be able to get references to a memory/reg/inst/etc as such that isn't referring to its underlying data).  Anyway, AFAICT we never actually claim a memory is or has a `vector<T>` or `T[len]` or whatever (in the spec, not in CIRCT).

I was hoping this would allow us to implement "mem taps" natively which for whatever reason are also per-element presently at least in practice re:generated annotations, and have the discussed weirdness re:packed/unpacked but this proposal maybe doesn't need to tackle every issue at once :laughing:.  Hence vector of references to sub-elements vs reference to vector of sub-elements seeming to be a good fit.  It seems like memories intentionally abstract away what the underlying "type" of their data is (no promises it's stored consecutively or any other way?), so maybe this is something to tackle later and disallow until this is worked on further.

Since I don't think I  was very clear, FWIW the language here is meant to specifically clarify that a probe of a memory is `Probe<T>[len]` not `Probe<T[len]>` (aka what I think we'd expect to get via the general rule: `Probe<decltype(memory.underlying_data_if_we_had_name_for_it_in_FIRRTL)>`).

Your point about worrying about unspecified lowering we happen to perform is well taken, thanks.  You're right that, at best, belongs in a section such as the compiler/implementation section.  And hopefully goes without saying if we want to give the data in a Memory a specific type, then `probe(mem)` should be `Probe<thattype>` if at all possible :+1: .  Not immediately sure what best short-term answer is, will circle back :+1: .

Reply:
I'm thinking there are things in the spec that are not there. I am good with the current language. 👍 Thanks for the detailed analysis.